src/avl_tree/set.rs

```diff
@@ -0,0 +1,400 @@
+            node.update_height();
+            node.rebalance();
+        }
+
+        true
+    }
+
+    pub fn take(&mut self, value: &T) -> Option<T> {
+        let mut prev_ptrs = Vec::<*mut AVLNode<T>>::new();
+        let mut current_tree = &mut self.root;
+        let mut target_value = None;
+
+        while let Some(current_node) = current_tree {
+            match current_node.value.cmp(&value) {
+                Ordering::Less => {
+                    prev_ptrs.push(&mut **current_node);
+                    current_tree = &mut current_node.right;
+                }
+                Ordering::Equal => {
+                    target_value = Some(&mut **current_node);
+                    break;
+                }
+                Ordering::Greater => {
+                    prev_ptrs.push(&mut **current_node);
+                    current_tree = &mut current_node.left;
+                }
+            }
+        }
+
+        // Target does not exist in tree
+        if target_value.is_none() {
+            return None;
+        }
+
+        let target_node = target_value.unwrap();
+
+        // 3 Cases: No children, left child, right child
+        let taken = if target_node.left.is_none() || target_node.right.is_none() {
+            // Left child
+            if let Some(left_node) = target_node.left.take() {
+                replace(target_node, *left_node).value
+            }
+            // Right child
+            else if let Some(right_node) = target_node.right.take() {
+                replace(target_node, *right_node).value
+            }
+            // No children
+            else if let Some(prev_ptr) = prev_ptrs.pop() {
+                let prev_node = unsafe { &mut *prev_ptr };
+
+                // Determine which node to take
+                let inner_value = if let Some(left_node) = prev_node.left.as_ref() {
+                    // Target was left node from parent
+                    if left_node.value == target_node.value {
+                        prev_node.left.take().unwrap().value
+                    }
+                    // Target was right node from parent
+                    else {
+                        prev_node.right.take().unwrap().value
+                    }
+                }
+                // Target was right all along, since left is None
+                else {
+                    prev_node.right.take().unwrap().value
+                };
+
+                prev_node.update_height();
+                prev_node.rebalance();
+
+                inner_value
+            }
+            // Fourth boring case: Tree is root
+            else {
+                self.root.take().unwrap().value
+            }
+        }
+        // Find Inorder-Successor
+        else {
+            AVLTreeSet::find_inorder_succesor(target_node)
+        };
+
+        // Update for every touched Node
+        for node_ptr in prev_ptrs.into_iter().rev() {
+            let node = unsafe { &mut *node_ptr };
```

Comment:
I guess you've thought this through and came to the conclusion that this unsafe block is the only way? Would be super duper n🍦 if we could remove it, so we can call the repo "safe" Rust, but it's not at all bad if this doesn't happen now.

Reply:
The problem is i don't know exactly why this is unsafe, but it would compile otherwise